### ISSUE TITLE
RBM/PretrainNetwork parameter flattening incorrect for backprop

### ISSUE THREAD

**Issue by Shima32** [deeplearning4j/deeplearning4j#780]
Alex Black comment on me problem : I think I know what's going on so RBMs have 3 types of parameters: weights, hidden bias, visible bias during pretraining we use all 3
during backprop we treat the RBM as a MLP and so just use two of the parameters (weights + hidden bias) however, based on the shapes of the input and the stack trace: it looks like network parameters (incorrectly) includes all 3 types of RBM parameters (instead of 2) whereas gradients (correctly) only include the 2 types

Stack Trace:
Oct 14, 2015 3:55:33 PM com.github.fommil.jni.JniLoader liberalLoad
INFO: successfully loaded C:\Users\shimam\AppData\Local\Temp\jniloader4238490470366948938netlib-native_ref-win-i686.dll
Oct 14, 2015 3:55:33 PM com.github.fommil.jni.JniLoader liberalLoad
INFO: successfully loaded C:\Users\shimam\AppData\Local\Temp\jniloader4751272666911484503netlib-native_system-win-i686.dll
o.d.n.m.MultiLayerNetwork - Training on layer 1 with 200 examples
o.d.o.s.BaseOptimizer - Objective function automatically set to minimize. Set stepFunction in neural net configuration to change default settings.
o.d.o.s.BaseOptimizer - Objective function automatically set to minimize. Set stepFunction in neural net configuration to change default settings.
o.d.o.l.ScoreIterationListener - Score at iteration 0 is 2.615318140341163
o.d.o.l.ScoreIterationListener - Score at iteration 20 is 2.680378213749647
o.d.o.l.ScoreIterationListener - Score at iteration 40 is 2.6104382911150457
o.d.o.l.ScoreIterationListener - Score at iteration 60 is 2.6004986231133937
o.d.o.l.ScoreIterationListener - Score at iteration 80 is 2.445558707113028
o.d.n.m.MultiLayerNetwork - Training on layer 2 with 200 examples
o.d.o.s.BaseOptimizer - Objective function automatically set to minimize. Set stepFunction in neural net configuration to change default settings.
o.d.o.s.BaseOptimizer - Objective function automatically set to minimize. Set stepFunction in neural net configuration to change default settings.
o.d.o.l.ScoreIterationListener - Score at iteration 0 is 5.407326913958072
o.d.o.l.ScoreIterationListener - Score at iteration 20 is 5.2723569441638
o.d.o.l.ScoreIterationListener - Score at iteration 40 is 5.372386981579781
o.d.o.l.ScoreIterationListener - Score at iteration 60 is 5.427417032880307
o.d.o.l.ScoreIterationListener - Score at iteration 80 is 5.1074470856671335
o.d.n.m.MultiLayerNetwork - Training on layer 3 with 200 examples
o.d.o.s.BaseOptimizer - Objective function automatically set to minimize. Set stepFunction in neural net configuration to change default settings.
o.d.o.s.BaseOptimizer - Objective function automatically set to minimize. Set stepFunction in neural net configuration to change default settings.
o.d.o.l.ScoreIterationListener - Score at iteration 0 is 4.832604580257892
o.d.o.l.ScoreIterationListener - Score at iteration 20 is 4.532549481212616
o.d.o.l.ScoreIterationListener - Score at iteration 40 is 4.522494414281368
o.d.o.l.ScoreIterationListener - Score at iteration 60 is 4.667439318916321
o.d.o.l.ScoreIterationListener - Score at iteration 80 is 4.642384226578474
o.d.n.m.MultiLayerNetwork - Training on layer 4 with 200 examples
o.d.o.s.BaseOptimizer - Objective function automatically set to minimize. Set stepFunction in neural net configuration to change default settings.
o.d.o.s.BaseOptimizer - Objective function automatically set to minimize. Set stepFunction in neural net configuration to change default settings.
o.d.o.l.ScoreIterationListener - Score at iteration 0 is 2.805648804940164
o.d.o.l.ScoreIterationListener - Score at iteration 20 is 2.77065380322659
o.d.o.l.ScoreIterationListener - Score at iteration 40 is 2.775658802371144
o.d.o.l.ScoreIterationListener - Score at iteration 60 is 2.8206674344427585
o.d.o.l.ScoreIterationListener - Score at iteration 80 is 2.8356824394673406
o.d.n.m.MultiLayerNetwork - Training on layer 5 with 200 examples
o.d.o.s.BaseOptimizer - Objective function automatically set to minimize. Set stepFunction in neural net configuration to change default settings.
o.d.o.s.BaseOptimizer - Objective function automatically set to minimize. Set stepFunction in neural net configuration to change default settings.
o.d.o.l.ScoreIterationListener - Score at iteration 0 is 0.46651955706471204
o.d.o.l.ScoreIterationListener - Score at iteration 20 is 0.5065245631858706
o.d.o.l.ScoreIterationListener - Score at iteration 40 is 0.561529569287777
o.d.o.l.ScoreIterationListener - Score at iteration 60 is 0.5615345752512813
o.d.o.l.ScoreIterationListener - Score at iteration 80 is 0.4915395813148022
o.d.n.m.MultiLayerNetwork - Training on layer 6 with 200 examples
o.d.o.s.BaseOptimizer - Objective function automatically set to minimize. Set stepFunction in neural net configuration to change default settings.
o.d.o.s.BaseOptimizer - Objective function automatically set to minimize. Set stepFunction in neural net configuration to change default settings.
o.d.o.l.ScoreIterationListener - Score at iteration 0 is 2.781080294088125
o.d.o.l.ScoreIterationListener - Score at iteration 20 is 2.7961003093643186
o.d.o.l.ScoreIterationListener - Score at iteration 40 is 2.671120326413393
o.d.o.l.ScoreIterationListener - Score at iteration 60 is 2.826148769022703
o.d.o.l.ScoreIterationListener - Score at iteration 80 is 2.636188815897465
o.d.n.m.MultiLayerNetwork - Training on layer 7 with 200 examples
o.d.o.s.BaseOptimizer - Objective function automatically set to minimize. Set stepFunction in neural net configuration to change default settings.
o.d.o.s.BaseOptimizer - Objective function automatically set to minimize. Set stepFunction in neural net configuration to change default settings.
o.d.o.l.ScoreIterationListener - Score at iteration 0 is 4.716687861729622
o.d.o.l.ScoreIterationListener - Score at iteration 20 is 4.531727921161175
o.d.o.l.ScoreIterationListener - Score at iteration 40 is 4.386767984737873
o.d.o.l.ScoreIterationListener - Score at iteration 60 is 4.666808057228089
o.d.o.l.ScoreIterationListener - Score at iteration 80 is 4.491848131956577
o.d.n.m.MultiLayerNetwork - Finetune phase
o.d.o.s.BaseOptimizer - Objective function automatically set to minimize. Set stepFunction in neural net configuration to change default settings.
o.d.o.l.ScoreIterationListener - Score at iteration 0 is 4.012883911132812
o.d.o.l.ScoreIterationListener - Score at iteration 20 is 4.00514404296875
o.d.o.l.ScoreIterationListener - Score at iteration 40 is 3.997624816894531
o.d.o.l.ScoreIterationListener - Score at iteration 60 is 3.9909393310546877
o.d.o.l.ScoreIterationListener - Score at iteration 80 is 3.984224548339844
o.d.o.s.BaseOptimizer - Objective function automatically set to minimize. Set stepFunction in neural net configuration to change default settings.
Exception in thread "main" java.lang.IllegalArgumentException: Shapes do not match: x.shape=[1, 422], y.shape=[1, 381]
	at org.nd4j.linalg.api.parallel.tasks.cpu.CPUTaskFactory.getTransformAction(CPUTaskFactory.java:89)
	at org.nd4j.linalg.api.ops.executioner.DefaultOpExecutioner.doTransformOp(DefaultOpExecutioner.java:409)
	at org.nd4j.linalg.api.ops.executioner.DefaultOpExecutioner.exec(DefaultOpExecutioner.java:62)
	at org.nd4j.linalg.api.ndarray.BaseNDArray.subi(BaseNDArray.java:2699)
	at org.nd4j.linalg.api.ndarray.BaseNDArray.subi(BaseNDArray.java:2680)
	at org.deeplearning4j.optimize.stepfunctions.NegativeGradientStepFunction.step(NegativeGradientStepFunction.java:37)
	at org.deeplearning4j.optimize.solvers.StochasticGradientDescent.optimize(StochasticGradientDescent.java:60)
	at org.deeplearning4j.optimize.Solver.optimize(Solver.java:52)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.fit(MultiLayerNetwork.java:1352)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.fit(MultiLayerNetwork.java:1382)
	at AutomatedTextSummarization.nuralNetwork.CreatingNN.main(CreatingNN.java:93)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:140)

Network configuration:
public class CreatingNN {
    private static Logger log = LoggerFactory.getLogger(CreatingNN.class);

    public static void main(String[] args) throws  Exception {
        RecordReader recordReader = new CSVRecordReader(0,",");
        recordReader.initialize(new FileSplit(new ClassPathResource("features&Index.txt").getFile()));
        //reader,label index,number of possible labels
        //with a DataSetIterator, fetching one or more new examples with each iteration,
        // and loading those examples into a DataSet object that neural nets can work with.
        // Calling next on the iterator advances it one step, and returns the new value it points to.
        Nd4j.MAX_SLICES_TO_PRINT = 10;
        Nd4j.MAX_ELEMENTS_PER_SLICE = 10;

        final int numRows = 4;
        final int numColumns = 1;
        final int numInputs = 4;
        int outputNum = 3;
        int iterations = 100;
        long seed = 123;
        int listenerFreq = iterations/5;

        log.info("Load data....");
        org.nd4j.linalg.dataset.api.iterator.DataSetIterator iterator = new RecordReaderDataSetIterator(recordReader,200,4,50);

        log.info("Build model....");
        MultiLayerConfiguration conf = new NeuralNetConfiguration.Builder()
                .seed(seed)
                .iterations(iterations)
                .learningRate(1e-6f)
                .optimizationAlgo(OptimizationAlgorithm.STOCHASTIC_GRADIENT_DESCENT)
                .l1(1e-1).regularization(true).l2(2e-4)
                .list(8)
                .layer(0, new RBM.Builder().nIn(numRows * numColumns).nOut(10).lossFunction(LossFunctions.LossFunction.RMSE_XENT).build())
                .layer(1, new RBM.Builder().nIn(10).nOut(8).lossFunction(LossFunctions.LossFunction.RMSE_XENT).build())
                .layer(2, new RBM.Builder().nIn(8).nOut(5).lossFunction(LossFunctions.LossFunction.RMSE_XENT).build())
                .layer(3, new RBM.Builder().nIn(5).nOut(1).lossFunction(LossFunctions.LossFunction.RMSE_XENT).build())//encoding stops
                .layer(4, new RBM.Builder().nIn(1).nOut(5).lossFunction(LossFunctions.LossFunction.RMSE_XENT).build())//decoding starts
                .layer(5, new RBM.Builder().nIn(5).nOut(8).lossFunction(LossFunctions.LossFunction.RMSE_XENT).build())
                .layer(6, new RBM.Builder().nIn(8).nOut(10).lossFunction(LossFunctions.LossFunction.RMSE_XENT).build())
                .layer(7, new OutputLayer.Builder(LossFunctions.LossFunction.RMSE_XENT).nIn(10).nOut(numRows * numColumns).build())
                .pretrain(true).backprop(true)

                .build();

        MultiLayerNetwork model = new MultiLayerNetwork(conf);
        model.init();

        model.setListeners(Arrays.asList((IterationListener) new ScoreIterationListener(listenerFreq)));

        log.info("Train model....");
      //  while(iterator.hasNext()) {
            DataSet next = iterator.next();
            next.normalizeZeroMeanZeroUnitVariance();
            model.fit(new DataSet(next.getFeatureMatrix(), next.getFeatureMatrix()));
      //  }
        INDArray w;
        log.info("Evaluate weights....");
      /*  for(org.deeplearning4j.nn.api.Layer layer : model.getLayers()) {
             w = layer.getParam(DefaultParamInitializer.WEIGHT_KEY);
            log.info("Weights: " + w);
            System.out.println(w);
        }*/

      //  INDArray output = model.output(test.getFeatureMatrix());
       // String predicted = output.getRow(i).toString().trim();

        INDArray layer4=model.getLayer(4).input();
        System.out.println(layer4);

    }

**Comment by AlexDBlack** [comment]
Here's the problem:
RBM extends BasePretrainNetwork, extends BaseLayer
RBM (pretrain networks in general) have 3 types of parameters: weight, hidden bias, visible bias

BaseLayer defines params() and setParams() to go to/from 1d vector, and BasePretrainNetwork/RBM does not override this
- BaseLayer.params() and setParams() uses Nd4j.toFlattened('f',params.values()). Hence in the case of RBM/pretrain, this includes all 3 types of parameters
- params()/setParams() methods are used for both pretraining and fine tuning
  - current implementation works for pretraining: gradients and flattened parameters include all 3 types of parameters
  - current implementation does not work for backprop: gradients (correctly) include only 2 types of parameters, whereas flattened parameters include all 3. Hence they are different lengths and we get the exception in the issue above
- Simply overriding BasePretrainNetwork.params(): if we fix it for backprop we break it for pretrain
  - setParams() is easy enough to fix (i.e., can work out easy enough whether we have 2 or 3 (backprop vs. pretrain) as inputs from length)

I need to think some more about what the best solution is here

**Comment by AlexDBlack** [comment]
Fixed.
https://github.com/deeplearning4j/deeplearning4j/pull/833
